metrics/connection_tracer.go

```diff
@@ -0,0 +1,201 @@
+package metrics
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/quic-go/quic-go"
+	"github.com/quic-go/quic-go/internal/qerr"
+	"github.com/quic-go/quic-go/logging"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	connStarted = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricNamespace,
+			Name:      "connections_started_total",
+			Help:      "Connections Started",
+		},
+		[]string{"dir"},
+	)
+	connClosed = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricNamespace,
+			Name:      "connections_closed_total",
+			Help:      "Connections Closed",
+		},
+		[]string{"dir", "reason"},
+	)
+	connDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: metricNamespace,
+			Name:      "connection_duration_seconds",
+			Help:      "Duration of a Connection",
+			Buckets:   prometheus.ExponentialBuckets(1.0/16, 2, 25), // up to 24 days
+		},
+		[]string{"dir"},
+	)
+	connHandshakeDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: metricNamespace,
+			Name:      "handshake_duration_seconds",
+			Help:      "Duration of the QUIC Handshake",
+			Buckets:   prometheus.ExponentialBuckets(0.001, 1.3, 35),
+		},
+		[]string{"dir"},
+	)
+)
+
+// DefaultTracer returns a callback that creates a metrics ConnectionTracer.
+// The ConnectionTracer returned can be set on the quic.Config for a new connection.
+// It should be reused across QUIC connections.
+func DefaultTracer() func(_ context.Context, p logging.Perspective, _ logging.ConnectionID) *logging.ConnectionTracer {
+	return DefaultTracerWithRegisterer(prometheus.DefaultRegisterer)
+}
+
+// DefaultTracerWithRegisterer returns a callback that creates a metrics ConnectionTracer
+// using a given Prometheus registerer.
+func DefaultTracerWithRegisterer(registerer prometheus.Registerer) func(_ context.Context, p logging.Perspective, _ logging.ConnectionID) *logging.ConnectionTracer {
```

Comment:
You can register the metrics here once instead of registering them in the callback. The current method also works but all but the first registration will return AlreadyRegistrationError so we can just register them upfront here.

Reply:
`NewClientConnectionTracerWithRegisterer` and `NewServerConnectionTracerWithRegisterer` are public functions though, so we'll need to register in these code paths as well (and catch the `AlreadyRegisteredError`s there).

Reply:
That's a good point.